xlive/Blam/Engine/rasterizer/dx9/rasterizer_dx9_screen_effect.cpp

```diff
@@ -1,8 +1,217 @@
+			&& !*global_effect_flag_get()
+			&& g_render_scenario_screen_effects)
+		{
+			if (cinematic_in_progress())
+			{
+				s_rasterizer_cinematics_globals* rasterizer_cinematic_globals = rasterizer_cinematic_screen_effect_update(NULL);
+				if (rasterizer_cinematic_globals)
+				{
+					if (rasterizer_cinematic_globals->field_3C && rasterizer_cinematic_globals->field_40 > 0.f)
+					{
+						rasterizer_dx9_render_depth_of_field(
+							rasterizer_cinematic_globals->field_44.lower,
+							rasterizer_cinematic_globals->field_44.upper,
+							rasterizer_cinematic_globals->field_40);
+					}
+
+					if (rasterizer_cinematic_globals->field_AC.lower > 0.f)
+					{
+						rasterizer_dx9_perf_event_begin("crossfade", NULL);
+						rasterizer_dx9_render_crossfade(
+							rasterizer_cinematic_globals->field_AC.lower,
+							rasterizer_cinematic_globals->field_AC.upper);
+						rasterizer_dx9_perf_event_end();
+					}
+				}
+			}
+		}
+
+		if (global_scenario->screen_effect_references.count > 0)
+		{
+			real32 x = -0.5f / width;
+			real32 y = -0.5f / height;
+
+			real32 mesh_data[16];
+			mesh_data[0] = x - 1.f;
+			mesh_data[1] = y + 1.f;
+			mesh_data[2] = 0.f;
+			mesh_data[3] = 0.f;
+
+			mesh_data[4] = x + 1.f;
+			mesh_data[5] = y + 1.f;
+			mesh_data[6] = 1.f;
+			mesh_data[7] = 0.f;
+
+			mesh_data[8] = x - 1.f;
+			mesh_data[9] = y - 1.f;
+			mesh_data[10] = 0.f;
+			mesh_data[11] = 1.f;
+
+			mesh_data[12] = x + 1.f;
+			mesh_data[13] = y - 1.f;
+			mesh_data[14] = 1.f;
+			mesh_data[15] = 1.f;
+
+			for (int32 i = 0; i < global_scenario->screen_effect_references.count; i++)
+			{
+				s_scenario_screen_effect_reference* screen_effect_reference = global_scenario->screen_effect_references[i];
+				if (screen_effect_reference->screen_effect.index != NONE)
+				{
+					real32 interpolator_result[4] = { 0.f };
+					if (screen_effect_reference->primary_scenario_interpolator_index != NONE)
+					{
+						interpolator_result[0] = scenario_interpolator_get_result(screen_effect_reference->primary_scenario_interpolator_index);
+					}
+
+					if (screen_effect_reference->secondary_scenario_interpolator_index != NONE)
+					{
+						interpolator_result[1] = scenario_interpolator_get_result(screen_effect_reference->secondary_scenario_interpolator_index);
+					}
+
+					if (interpolator_result[0] != 0.f || interpolator_result[1] != 0.f)
+					{
+						target = rasterizer_screen_effects_render(
+							screen_effect_reference->screen_effect.index,
+							interpolator_result,
+							2,
+							2,
+							mesh_data);
+					}
+				}
+			}
+		}
+	}
+	
+
+	if (lens_flare_occlusion_test)
+	{
+		target = rasterizer_suns_glow_draw(target);
+	}
+
+	rasterizer_dx9_perf_event_begin("gamma_brightness", NULL);
+	rasterizer_dx9_apply_gamma_and_brightness(target);
+	rasterizer_dx9_perf_event_end();
+
+	global_d3d_device->StretchRect(global_d3d_surface_render_primary_get(), NULL, global_d3d_surface_render_resolved_get(), NULL, D3DTEXF_NONE);
```

Comment:
I don't understand the extra copy here

Reply:
The copy was originally done at the start of the function.
I moved it to the bottom once the primary render surface had the sun effect written to it.

Reply:
but why are you copying it to the resolved surface if the resolved surface gets copied in the back-buffer during `rasterizer_dx9_apply_gamma_and_brightness()`?

Reply:
It copies it from target_25 to the back-buffer in rasterizer_dx9_apply_gamma_and_brightness?

Reply:
it copies it from target 37, which is the "resolved" surface

Reply:
I debugged it and it's copying it from 25.

Reply:
yeah, 25 hex..., which is 37 in decimal